### PR TITLE
Add split_to_map function to string operations

### PR DESCRIPTION
**Pull request by monkey-mas**
ref. issue #2544 
This commit adds a function which is expected to work almost the same as hive's str_to_map() function that generates a map based on input string and two delimiters.

Additionally, we might need to discuss a proper function name.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 1]
```diff
@@ -359,6 +362,49 @@ public static Slice splitPart(@SqlType(StandardTypes.VARCHAR) Slice string, @Sql
...
+    {
+        checkCondition(delimiter1.length() > 0, INVALID_FUNCTION_ARGUMENT, "The delimiter1 may not be the empty string");
+        checkCondition(delimiter2.length() > 0, INVALID_FUNCTION_ARGUMENT, "The delimiter2 may not be the empty string");
+
+        Map<Slice, Slice> map = new HashMap<>();
+        int index = 0;
+        while (index < string.length()) {
+            int splitIndex = string.indexOf(delimiter1, index);
+            if (splitIndex < 0) { // Cannot be split by delimiter1
+                break;
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Instead of duplicating the logic inside and outside of the loop, write
```
if (splitIndex >= 0) {
  kvPair = string.slice(index, string.length() - index);
}
else {
  kvPair = string.slice(index, splitIndex - index);
}
```

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java [hunk 2]
```diff
@@ -238,6 +240,39 @@ public void testSplit()
...
 
     @Test
+    public void testSplitToMap()
+    {
+        ImmutableMap<String, String> abc   = ImmutableMap.of("a", "123", "b", ".4", "c", "");
+        ImmutableMap<String, String> ab    = ImmutableMap.of("a", "123", "b", ".4");
+        ImmutableMap<String, String> ac    = ImmutableMap.of("a", "123", "c", "");
+        ImmutableMap<String, String> bc    = ImmutableMap.of("b", ".4", "c", "");
+        ImmutableMap<String, String> a     = ImmutableMap.of("a", "123,b=.4,c=");
+        ImmutableMap<String, String> uabc  = ImmutableMap.of("\u0061", "\u0031\u0032\u0033", "\u0062", "\u002e\u0034", "\u0063", "");
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: 0020-007f are ascii. I recommend using 4E00-4EFF instead.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 3]
```diff
@@ -359,6 +362,49 @@ public static Slice splitPart(@SqlType(StandardTypes.VARCHAR) Slice string, @Sql
...
+            }
+            // Extract key-value pair based on current index
+            // then add the pair if it can be split by delimiter2
+            Slice kvPair = string.slice(index, splitIndex - index);
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: We don't use abbreviations in Presto code base. Call it `keyValuePair` instead.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java [hunk 4]
```diff
@@ -238,6 +240,39 @@ public void testSplit()
     }
 
     @Test
+    public void testSplitToMap()
+    {
+        ImmutableMap<String, String> abc   = ImmutableMap.of("a", "123", "b", ".4", "c", "");
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: We don't align `=`s

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 5]
```diff
@@ -359,6 +362,49 @@ public static Slice splitPart(@SqlType(StandardTypes.VARCHAR) Slice string, @Sql
...
+            Slice key    = splitPart(kvPair, delimiter2, 1);
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Use `indexOf` and `slice` as you did for `delimiter1`. `splitPart` does quite a lot of magic under the hood and is designed more toward use directly from SQL.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 6]
```diff
@@ -359,6 +362,49 @@ public static Slice splitPart(@SqlType(StandardTypes.VARCHAR) Slice string, @Sql
         return null;
     }
 
+    @Description("creates a map using delimiter1 and delimiter2")
+    @ScalarFunction
+    @SqlType("map<varchar,varchar>")
+    public static Slice splitToMap(@SqlType(StandardTypes.VARCHAR) Slice string, @SqlType(StandardTypes.VARCHAR) Slice delimiter1, @SqlType(StandardTypes.VARCHAR) Slice delimiter2)
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Call the arguments `entryDelimiter` and `keyValueDelimiter` instead of `delimiter1` and `delimiter2`.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 7]
```diff
@@ -359,6 +362,49 @@ public static Slice splitPart(@SqlType(StandardTypes.VARCHAR) Slice string, @Sql
...
+    {
+        checkCondition(delimiter1.length() > 0, INVALID_FUNCTION_ARGUMENT, "The delimiter1 may not be the empty string");
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Change the variable names in these strings accordingly.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 8]
```diff
@@ -359,6 +362,49 @@ public static Slice splitPart(@SqlType(StandardTypes.VARCHAR) Slice string, @Sql
...
+        checkCondition(delimiter2.length() > 0, INVALID_FUNCTION_ARGUMENT, "The delimiter2 may not be the empty string");
+
+        Map<Slice, Slice> map = new HashMap<>();
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Don't use `HashMap` as intermediate representation. Construct a `BlockBuilder` using `VARCHAR.createBlockBuilder`. And then use `VARCHAR.writeSlice` to append key1, value1, key2, value2, etc into the Block. At the end, return the result you got from `.build()`. By the way, the return type of functions that return SQL map is changed to `Block` as a result of recent major refactor.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 5 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: In addition, when the user do `key=va=lue`, let's be lenient and construct a map of `{'key': 'va=lue'}`.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java [hunk 9]
```diff
@@ -238,6 +240,39 @@ public void testSplit()
...
+        ImmutableMap<String, String> bc    = ImmutableMap.of("b", ".4", "c", "");
+        ImmutableMap<String, String> a     = ImmutableMap.of("a", "123,b=.4,c=");
+        ImmutableMap<String, String> uabc  = ImmutableMap.of("\u0061", "\u0031\u0032\u0033", "\u0062", "\u002e\u0034", "\u0063", "");
+        ImmutableMap<String, String> uab   = ImmutableMap.of("\u0061", "\u0031\u0032\u0033", "\u0062", "\u002e\u0034");
+        ImmutableMap<String, String> uac   = ImmutableMap.of("\u0061", "\u0031\u0032\u0033", "\u0063", "");
+        ImmutableMap<String, String> ubc   = ImmutableMap.of("\u0062", "\u002e\u0034", "\u0063", "");
+        ImmutableMap<String, String> ua    = ImmutableMap.of("\u0061", "\u0031\u0032\u0033\u002c\u0062\u003d\u002e\u0034\u002c\u0063\u003d");
+        ImmutableMap<String, String> empty = ImmutableMap.of();
+
+        assertFunction("SPLIT_TO_MAP('a=123,b=.4,c=', ',', '=')", new MapType(VARCHAR, VARCHAR), abc);
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: You tested empty string as value here. Let's also test empty string as key.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java [hunk 10]
```diff
@@ -359,6 +362,49 @@ public static Slice splitPart(@SqlType(StandardTypes.VARCHAR) Slice string, @Sql
...
+        while (index < string.length()) {
+            int splitIndex = string.indexOf(delimiter1, index);
+            if (splitIndex < 0) { // Cannot be split by delimiter1
+                break;
+            }
+            // Extract key-value pair based on current index
+            // then add the pair if it can be split by delimiter2
+            Slice kvPair = string.slice(index, splitIndex - index);
+            Slice key    = splitPart(kvPair, delimiter2, 1);
+            if (key != null) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: What do you think about throwing an exception in this case?

**Comment by haozhun** [comment]
Ping me after you rebase and resolve the comments. Thanks.

**Comment by haozhun** [comment]
This function is useful. However, it doesn't look like the original author is still working on it.

If any one is interested, feel free to submit a new pull request.

**Comment by monkey-mas** [comment]
Hi @haozhun,
Sorry for my late response. I rebased and modified my PR based on your suggestions. Please review it if you have time. Thanks.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 1 above)
Comment (monkey-mas) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: I suppose you meant the opposite conditions?
```
if (splitIndex >= 0) {
    keyValuePair = string.slice(index, splitIndex - index);
}
else {
    keyValuePair = string.slice(index, string.length() - index);
}
```
Anyway, thanks for pointing it out :)

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java (hunk 2 above)
Comment (monkey-mas) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: Sure, done :)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 5 above)
Comment (monkey-mas) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Done :)

> In addition, when the user do key=va=lue, let's be lenient and construct a map of {'key': 'va=lue'}.

Sure, that's what I assumed :)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java (hunk 6 above)
Comment (monkey-mas) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/StringFunctions.java]: Strongly agree. That's much better for us to understand :)

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java (hunk 9 above)
Comment (monkey-mas) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestStringFunctions.java]: Agree. Done!

**Comment by cberner** [comment]
I think we should make some changes in the way a couple edge cases work:

1) an empty value should not be skipped, it should be empty string instead. That is `split_to_map('1=', ';', '=')` is equal to `map(array['1'], array[''])`
2) duplicate keys are not allowed. That is `split_to_map('1=a;1=b', ';', '=')` should throw `INVALID_FUNCTION_ARGUMENT` error code
3) entry delimiter and key-value delimiter must not be equal. That is `split_to_map('1=', '=', '=')` should throw `INVALID_FUNCTION_ARGUMENT` error code
4) key-value delimiter must appear exactly once in each entry. That is `split_to_map('1==', ';', '=')` and `split_to_map('1', ';', '=')` should throw `INVALID_FUNCTION_ARGUMENT` error code

**Comment by monkey-mas** [comment]
Hi @cberner,

Thanks for your advice. I rebased and applied your ideas. I'll squash all the PRs into one after everything looks fine to you.

> 1) an empty value should not be skipped, it should be empty string instead. That is split_to_map('1=', ';', '=') is equal to map(array['1'], array[''])

I think this is already satisfied according to this [test](https://github.com/prestodb/presto/pull/3142/commits/6ba553feb7d4f8950d3a1ecacf3073d971f8a172#diff-120bfcb8f8275fb3e7d01554805f7b4cR252), right?
Do you think we could also allow an [empty key](https://github.com/prestodb/presto/pull/3142/commits/6ba553feb7d4f8950d3a1ecacf3073d971f8a172#diff-120bfcb8f8275fb3e7d01554805f7b4cR264) as well or not?

> 2) duplicate keys are not allowed. That is split_to_map('1=a;1=b', ';', '=') should throw INVALID_FUNCTION_ARGUMENT error code

Sure, [done](https://github.com/prestodb/presto/pull/3142/commits/cf2e09a5e288fd29437c938f7dbdf6a95d258c2e). I used Map to check if there're duplicate keys. If this is not fine, please let me know the better way :)

> 3) entry delimiter and key-value delimiter must not be equal. That is split_to_map('1=', '=', '=') should throw INVALID_FUNCTION_ARGUMENT error code

Agree. I've done [it](https://github.com/prestodb/presto/pull/3142/commits/965c9b53cbc3c5fe3fb217e2c6fbd25e5ed15d02).

> 4) key-value delimiter must appear exactly once in each entry. That is split_to_map('1==', ';', '=') and split_to_map('1', ';', '=') should throw INVALID_FUNCTION_ARGUMENT error code

I think this requirement is necessary to solve the ambiguity of "which word is used(deleted) to split key-value entry"? Then I agree :) [commit](https://github.com/prestodb/presto/pull/3142/commits/afa3f762ce6cdd81da212dc2e946f71628862fd3)

**Comment by cberner** [comment]
Merged. Thanks!

**Comment by monkey-mas** [comment]
Thanks for your help!
I think you might close [https://github.com/prestodb/presto/issues/2544](https://github.com/prestodb/presto/issues/2544) as well?
